### PR TITLE
GH-3042 no volatile atomic in iterators that are not threadsafe

### PR DESCRIPTION
**Pull request by JervenBolleman**
GitHub issue resolved: #3042 

Briefly describe the changes proposed in this PR:

Remove volatile and synchronized in iterators that are inherently not threadsafe.

----
PR Author Checklist (see the [contributor guidelines](https://github.com/eclipse/rdf4j/blob/main/.github/CONTRIBUTING.md) for more details):

 - [x] my pull request is [self-contained](https://rdf4j.org/documentation/developer/merge-strategy/#self-contained-changes-pull-requests-and-commits)
 - [ ] I've added tests for the changes I made
 - [x] I've applied [code formatting](https://github.com/eclipse/rdf4j/blob/main/.github/CONTRIBUTING.md#code-formatting) (you can use `mvn process-resources` to format from the command line)
 - [ ] I've [squashed](https://rdf4j.org/documentation/developer/squashing) my commits where necessary 
 - [x] every commit message starts with the issue number (GH-xxxx) followed by a meaningful description of the change

### CODE REVIEW

**Comment by JervenBolleman** [comment]
This is removing some support for thread safety in iterators. Unfortunately, iterators are inherently unsafe to use from multiple threads as hasNext/next are not an "atomic" operation but two independent one. So these should be removed to avoid people thinking these are threadsafe.

File: core/queryalgebra/evaluation/src/test/java/org/eclipse/rdf4j/query/algebra/evaluation/impl/FilterOptimizerTest.java [hunk 1]
```diff
@@ -75,8 +77,9 @@ public void deMerge() {
 		Filter spo = new Filter(new StatementPattern(s, p, o), oSmallerThanTwo);
 		Compare o2SmallerThanFour = new Compare(o2, four, CompareOp.LT);
 		Filter spo2 = new Filter(new StatementPattern(s, p, o2), o2SmallerThanFour);
-		TupleExpr expected = new Projection(new Join(spo, spo2), new ProjectionElemList(new ProjectionElem("s"),
-				new ProjectionElem("p"), new ProjectionElem("o"), new ProjectionElem("o2")));
+		TupleExpr expected = new QueryRoot(
```
Comment (jeenbroekstra) [inline core/queryalgebra/evaluation/src/test/java/org/eclipse/rdf4j/query/algebra/evaluation/impl/FilterOptimizerTest.java:80]: This change does not seem related to the thread-safety changes - accidentally left in?

Comment (JervenBolleman) [reply core/queryalgebra/evaluation/src/test/java/org/eclipse/rdf4j/query/algebra/evaluation/impl/FilterOptimizerTest.java:80]: Yes the first commit in the series should not be there. I will try to fix and force push this.

**Review by jeenbroekstra — APPROVED** [review]
All looks good to me, thanks @JervenBolleman !

**Comment by JervenBolleman** [comment]
@jeenbroekstra that means I can hit the merge button now?

**Comment by jeenbroekstra** [comment]
Yep go for it!
